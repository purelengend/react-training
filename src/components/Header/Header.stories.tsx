import type { Meta, StoryObj } from '@storybook/react';
import Header from './index';
import './Header.module.css';
import layoutStyles from '../../layout/layout.module.css';
const meta: Meta<typeof Header> = {
  title: 'Example/Header',
  component: Header,
  decorators: [
    Story => (
      <div className={layoutStyles.container}>
        <Story />
      </div>
    )
  ]
};

type Story = StoryObj<typeof Header>;

export const HeaderExample: Story = {};
export default meta;
