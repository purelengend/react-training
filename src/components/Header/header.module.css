.header-container {
  padding: var(--padding-md) var(--padding-xl) var(--padding-sm);
  color: var(--text-primary);
}

.header-main-wrapper {
  justify-content: space-between;
}

.header-brand {
  font-size: var(--font-size-lg);
  font-weight: var(--font-weight-bold);
  line-height: var(--line-height-lg);
}

.header-brand:active {
  color: var(--primary-color);
}

.search-form {
  background-color: var(--base-form-bg);
  padding: var(--padding-xs);
  width: 342px;
  border: 1px solid var(--base-dark-line);
  border-radius: var(--border-primary);
  gap: 8px;
}

.search-input {
  background: transparent;
  border: none;
  outline: none;
  align-self: end;
  font-size: var(--font-size-sm);
  color: var(--text-light);
  line-height: var(--line-height-lg);
}

.search-input::placeholder {
  font-size: var(--font-size-sm);
  color: var(--text-light);
  line-height: var(--line-height-lg);
}

.header-sub-wrapper {
  justify-content: end;
  margin-top: 38px;
}
