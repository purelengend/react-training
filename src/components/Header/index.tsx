import searchIcon from '@assets/icons/search-icon.svg';
import { InputField } from '@components/common/InputField';
import { Select, SelectOptionProps } from '@components/common/Select';
import headerStyles from '@components/Header/header.module.css';
import { FILTER_ATTRIBUTE } from '@constants/filter';
import { ModalContext } from '@context/modal';
import { UrlContext } from '@context/url';
import { useDebounce } from '@hooks/useDebounce';
import useFood from '@hooks/useFood';
import { useCallback, useContext, useEffect, useMemo, useState } from 'react';

const Header = () => {
  const { path, sortFilter, setSortFilter, searchName, setSearchName } =
    useContext(UrlContext);

  const { setLoadingShowUp } = useContext(ModalContext);

  const { refetch, isRefetching } = useFood();

  const [searchText, setSearchText] = useState('');

  const debouncedText = useDebounce(searchText.replace(/\s+/g, ' ').trim()); // remove extra spaces of the search text

  // This effect is to refetch the list food based on path and search name
  useEffect(() => {
    refetch();
  }, [refetch, path, searchName]);

  useEffect(
    () => setLoadingShowUp(isRefetching),
    [isRefetching, setLoadingShowUp]
  );

  useEffect(() => setSearchName(debouncedText), [debouncedText, setSearchName]);

  const onRefresh = useCallback(
    (e: React.MouseEvent<HTMLAnchorElement>): void => {
      e.preventDefault();

      window.location.reload();
    },
    []
  );

  const selectOptions: SelectOptionProps[] = useMemo(
    () => [
      {
        value: undefined,
        disabled: true,
        label: 'Sort by price'
      },
      {
        value: FILTER_ATTRIBUTE.DEFAULT,
        disabled: false,
        label: 'Default'
      },
      {
        value: FILTER_ATTRIBUTE.ASCENDING,
        disabled: false,
        label: 'Ascending'
      },
      {
        value: FILTER_ATTRIBUTE.DESCENDING,
        disabled: false,
        label: 'Descending'
      }
    ],
    []
  );

  const onChangeSearchInput = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => setSearchText(e.target.value),
    []
  );

  const onChangeSelectOption = useCallback(
    (e: React.ChangeEvent<HTMLSelectElement>) => {
      setSortFilter(e.target.value);
    },
    [setSortFilter]
  );

  const onSubmitSearchForm = useCallback(
    (e: React.FormEvent<HTMLFormElement>) => e.preventDefault(),
    []
  );
  return (
    <header className={headerStyles['header-container']}>
      <div className={`d-flex ${headerStyles['header-main-wrapper']}`}>
        <a href="#" onClick={onRefresh}>
          <h1 id="header" className={headerStyles['header-brand']}>
            Foods Management
          </h1>
        </a>
        <form
          onSubmit={onSubmitSearchForm}
          className={`d-flex ${headerStyles['search-form']}`}
        >
          <InputField
            htmlFor="search"
            type="text"
            name="search"
            inputClass={headerStyles['search-input']}
            placeholder="Search for food, coffee, etc.."
            value={searchText}
            onChange={onChangeSearchInput}
            label={
              <img
                width="16"
                height="16"
                src={searchIcon}
                alt="Search Icon"
                className={headerStyles['primary-icon']}
              />
            }
          />
        </form>
      </div>
      <div className={`d-flex ${headerStyles['header-sub-wrapper']}`}>
        <Select
          value={sortFilter}
          onChange={onChangeSelectOption}
          selectOptions={selectOptions}
        />
      </div>
    </header>
  );
};

export default Header;
