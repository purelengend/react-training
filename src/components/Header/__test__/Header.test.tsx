import { customRender, customRenderer, screen } from '@src/test/test-utils';
import userEvent from '@testing-library/user-event';

import Header from '..';

// This script is to prevent error messages when mocking anchor tag onclick function
const listeners = window._virtualConsole.listeners('jsdomError');
const originalListener = listeners && listeners[0];

window._virtualConsole.removeAllListeners('jsdomError');

// Add a new listener to swallow JSDOM errors that ordinate from clicks on anchor tags.
window._virtualConsole.addListener('jsdomError', error => {
  if (
    error.type !== 'not implemented' &&
    error.message !== 'Not implemented: navigation (except hash changes)' &&
    originalListener
  ) {
    originalListener(error);
  }

  // swallow error
});

describe('Header test case', () => {
  it('should render correctly', () => {
    const mockHeader = customRenderer(<Header />);

    expect(mockHeader.toJSON()).toMatchSnapshot();
  });

  it('should invoke refresh function when click the heading', async () => {
    customRender(<Header />);

    const mockOnRefresh = vi.fn();

    const heading = screen.getByRole('link', {
      name: /foods management/i
    });

    heading.onclick = mockOnRefresh;

    await userEvent.click(heading);

    expect(mockOnRefresh).toHaveBeenCalled();
  });

  it('should invoke onChange function when select options', async () => {
    customRender(<Header />);

    const mockOnChangeSelection = vi.fn();

    const select = screen.getByRole('combobox', {
      name: /sort/i
    });

    select.onchange = mockOnChangeSelection;

    const selectOption = screen.getByRole('option', { name: 'Ascending' });

    await userEvent.selectOptions(select, selectOption);

    expect(mockOnChangeSelection).toHaveBeenCalled();
  });
});
