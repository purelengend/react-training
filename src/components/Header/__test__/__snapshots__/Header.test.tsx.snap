// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header test case > should render correctly 1`] = `
<header
  className="_header-container_d33141"
>
  <div
    className="d-flex _header-main-wrapper_d33141"
  >
    <div>
      <a
        href="#"
        onClick={[Function]}
      >
        <h1
          className="_header-brand_d33141"
          id="header"
        >
          Foods Management
        </h1>
      </a>
    </div>
    <div
      className="d-flex _search-form_d33141"
    >
      <label
        htmlFor="search"
      >
        <img
          alt="Search Icon"
          height="16"
          src="/src/assets/icons/search-icon.svg"
          width="16"
        />
      </label>
      <input
        className="_search-input_d33141"
        id="search"
        name="search"
        onChange={[Function]}
        placeholder="Search for food, coffee, etc.."
        step="any"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="d-flex _header-sub-wrapper_d33141"
  >
    <select
      aria-label="sort"
      className="_sort-select_e72513"
      id="sort"
      onChange={[Function]}
      value="orderby=createdAt&order=desc"
    >
      <option
        className="_select-item_1f6fcf"
        disabled={true}
      >
        Sort by price
      </option>
      <option
        className="_select-item_1f6fcf"
        disabled={false}
        value="orderby=createdAt&order=desc"
      >
        Default
      </option>
      <option
        className="_select-item_1f6fcf"
        disabled={false}
        value="orderby=price"
      >
        Ascending
      </option>
      <option
        className="_select-item_1f6fcf"
        disabled={false}
        value="orderby=price&order=desc"
      >
        Descending
      </option>
    </select>
  </div>
</header>
`;
